charabia/src/segmenter/camel_case.rs

```diff
@@ -0,0 +1,92 @@
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+pub(crate) trait CamelCaseSegmentation {
+    /// Returns an iterator over substrings of `self` separated on camelCase boundaries.
+    /// For instance, "camelCase" is split into ["camel", "Case"].
+    /// A camelCase boundary constitutes a lowercase letter directly followed by an uppercase letter
+    /// where lower and uppercase letters are defined by the corresponding Unicode General Categories.
+    fn split_camel_case_bounds(&self) -> CamelCaseParts;
+}
+
+pub(crate) struct CamelCaseParts<'t> {
+    state: State<'t>,
+}
+
+enum State<'t> {
+    InProgress { remainder: &'t str },
+    Exhausted,
+}
+
+impl CamelCaseSegmentation for str {
+    fn split_camel_case_bounds(&self) -> CamelCaseParts {
+        CamelCaseParts { state: State::InProgress { remainder: self } }
+    }
```

Comment:
An alternative to implementing the iterator manually would be to use some of the existing iterator utilities, perhaps [`iter::from_fn()`](https://doc.rust-lang.org/std/iter/fn.from_fn.html):
```rust
fn split_camel_case_bounds(&self) -> impl Iterator<Item = &str> + '_ {
    let mut remainder = self;

    iter::from_fn(move || {
        if reminder.is_empty() {
            return None;
        }
        
        match reminder.find(...)
            ...
            return Some(remaidner[..boundary]);
    })
}
```

I have not tried, but this simplified example compiles: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c19ba3e384445a3c2cf054625134949f

**Caveat**: Rust currently doesn't allow "[Return position impl Trait in traits](https://rust-lang.github.io/impl-trait-initiative/RFCs/rpit-in-traits.html#return-position-impl-trait-in-traits-1)" (this is just a very draft RFC), so one cannot have the `-> impl Iterator<Item = &str> + '_` signature in `CamelCaseSegmentation`.

And I fear one cannot even use the specific [`FromFn`](https://doc.rust-lang.org/std/iter/struct.FromFn.html) struct as its generic closure argument `F` cannot be named.. :thinking:

Reply:
I did try this over [here](https://github.com/goodhoko/charabia/commit/8f2173eaff6582aed5c745e58ed07bae3ed01013) and, to be honest, I think it doesn't really make the code any more simple or readable.

BTW: I know it's not the point of this suggestion, but using the bare remainder to track the state of the iterator is not merely enough. When the remainder is empty you may either want to yield and empty string or `None` depending on whether you were given an empty string to begin with or not. That's because splitting `""` should result in `[""]` and not `[]`.

Reply:
Right, I see. I didn't realize the complication with an empty string.